python/python/lance/progress.py

```diff
@@ -148,17 +148,19 @@ def begin(
             The multipart id to upload this fragment to cloud storage.
         """
 
-        self._fs.create_dir(self._base_path)
+        self._fs.create_dir(self._base_path, recursive=True)
+
         with self._fs.open_output_stream(self._in_progress_path(fragment)) as out:
             progress_data = {
                 "fragment_id": fragment.id,
                 "multipart_id": multipart_id if multipart_id else "",
                 "metadata": self._metadata,
             }
-            json.dump(progress_data, out)
-        with self._fs.open_input_stream(self._fragment_file(fragment)) as out:
-            out.write(fragment.to_json()).encode("utf-8")
+            out.write(json.dumps(progress_data).encode("utf-8"))
```

Comment:
is the change for the case where metadata contains non-string bytes?

Reply:
There are two changes here:

1. Wrapping `fragment.to_json()` with `json.dumps()`. This is because `to_json()` returns `Dict[str, Any]`, not `str`.
2. Moving the `.encode("utf-8")` to be called on the JSON string, and not on the output of `out.write()`. We were passing in string to `out.write()` and it accepts bytes only.